include/zephyr/logging/log.h

```diff
@@ -89,13 +91,455 @@ extern "C" {
+ * @param _str     Persistent, raw string.
+ */
+#define _LOG_HEXDUMP_RATELIMIT_CORE(_level, _rate_ms, _data, _length, _str)	\
+	do {									\
+		static atomic_t __last_log_time ;		\
+		static atomic_t __skipped_count ;		\
+		uint32_t __now = k_uptime_get_32();				\
+		uint32_t __last = atomic_get(&__last_log_time);		\
+		uint32_t __diff = __now - __last;				\
+		if (unlikely(__diff >= (_rate_ms))) {			\
+			if (atomic_cas(&__last_log_time, __last, __now)) {	\
+				uint32_t __skipped = atomic_clear(&__skipped_count);	\
+				if (__skipped > 0) {				\
+					Z_LOG(_level, "Skipped %d hexdump messages", __skipped); \
+				}						\
+				Z_LOG_HEXDUMP(_level, _data, _length, _str);	\
+			} else {						\
+				atomic_inc(&__skipped_count);			\
+			}							\
+		} else {							\
+			atomic_inc(&__skipped_count);				\
+		}								\
+	} while (0)
+
+/**
+ * @brief Rate-limited logging macros
+ *
+ * @details These macros provide rate-limited logging functionality to prevent
+ * log flooding when messages are generated frequently. Each macro ensures that
+ * log messages are not output more frequently than a specified interval.
+ * Rate limiting is per-macro-call-site, meaning each unique call has its own
+ * independent rate limit.
+ *
+ * The macros use atomic operations to ensure thread safety in multi-threaded
+ * environments. Only one thread can successfully log a message within the
+ * specified rate limit period.
+ *
+ * @see CONFIG_LOG_RATELIMIT_INTERVAL_MS
+ */
+
+/**
+ * @brief Core rate-limited logging macro with level parameter
+ *
+ * @details Internal macro that provides rate-limited logging functionality
+ * with configurable log level. Uses atomic operations to ensure thread safety
+ * in multi-threaded environments.
+ *
+ * @param _level Log level (LOG_LEVEL_ERR, LOG_LEVEL_WRN, etc.)
+ * @param _rate_ms Minimum interval in milliseconds between log messages.
+ * @param ... Arguments to pass to the logging function.
+ */
+#define _LOG_RATELIMIT_LVL(_level, _rate_ms, ...) \
+	do { \
+		if (IS_ENABLED(CONFIG_LOG_RATELIMIT)) { \
+			_LOG_RATELIMIT_CORE(_level, _rate_ms, __VA_ARGS__); \
+		} else if (IS_ENABLED(CONFIG_LOG_RATELIMIT_FALLBACK_DROP)) { \
+			(void)0; \
+		} else { \
+			Z_LOG(_level, __VA_ARGS__); \
+		} \
+	} while (0)
+
+/**
+ * @brief Writes a WARNING level message to the log with rate limiting.
+ *
+ * @details It's meant for situations that warrant investigation but could clutter
+ * the logs if output too frequently. The message will be logged at most once
+ * per default interval (see CONFIG_LOG_RATELIMIT_INTERVAL_MS).
+ *
+ * @param ... A string optionally containing printk valid conversion specifier,
+ * followed by as many values as specifiers.
+ */
+#define LOG_WRN_RATELIMIT(...) \
+	COND_CODE_1(CONFIG_LOG_RATELIMIT, \
+		(_LOG_RATELIMIT_LVL(LOG_LEVEL_WRN, CONFIG_LOG_RATELIMIT_INTERVAL_MS, \
+				   __VA_ARGS__)), \
+		(COND_CODE_1(CONFIG_LOG_RATELIMIT_FALLBACK_DROP, \
+			((void)0), \
+			(LOG_WRN(__VA_ARGS__)))))
```

Comment:
I fail to understand the need for `COND_CODE_1` here? Is it because of an undefined `CONFIG_LOG_RATELIMIT_INTERVAL_MS`?

How about:
```c
#ifdef CONFIG_LOG_RATELIMIT
#define LOG_RATELIMIT_INTERVAL_MS CONFIG_LOG_RATELIMIT_INTERVAL_MS
#else
#define LOG_RATELIMIT_INTERVAL_MS 0
#endif

// ...
#define LOG_WRN_RATELIMIT(...) \
	_LOG_RATELIMIT_LVL(LOG_LEVEL_WRN, LOG_RATELIMIT_INTERVAL_MS, __VA_ARGS__)

Reply:
>  Is it because of an undefined CONFIG_LOG_RATELIMIT_INTERVAL_MS?

yes.

> How about:

Yeah, we can try this as well,